### PR TITLE
Windowstate for windows

### PR DESCRIPTION
**Pull request by jesicasusanto**
Address issue #75 
This PR implements getting the window state of an active window with pywinauto library. Changes I've made :
- implement `get_window_state()` to get the data of the active window elements properties in the form of dictionary. Each window state contains a dictionary with keys `["title", "left", "top", "width", "height", "meta", "data", "window_id", "children"]`. 

Example of recording visualization : 
[recording-example.pdf](https://github.com/MLDSAI/OpenAdapt/files/11730909/recording-example.pdf)

### CODE REVIEW

**Comment by jesicasusanto** [comment]
Update :
I got an error from `get_descendants_info(window)` when running main()  :
`  File "C:\Users\jesic\PycharmProjects\PAT\openadapt\window\_windows.py", line 55, in get_descendants_info
    super(ComboBoxWrapper, self).expand()
  File "C:\Users\jesic\PycharmProjects\PAT\openadapt\.venv\lib\site-packages\pywinauto\controls\uiawrapper.py", line 527, in expand
    self.iface_expand_collapse.Expand()
_ctypes.COMError: (-2147220992, None, (None, None, None, 0, None))`

File: openadapt/window/_windows.py [hunk 1]
```diff
@@ -1,30 +1,72 @@
...
     return state
 
+def get_active_window_meta(active_window) :
+    if not active_window:
+        logger.warning(f"{active_window=}")
+        return None
+    logger.info(f"{active_window.get_properties()}=")
+    return active_window.get_properties()
+
+def get_active_element_state(x, y):
```
Comment (abrichr) [inline openadapt/window/_windows.py:32]: Two newlines before functions please 🙏

File: openadapt/window/_windows.py [hunk 2]
```diff
@@ -1,30 +1,72 @@
...
+    state = get_descendants_info(active_window)
+    logger.info(f"{state=}")
+    #state = window.dump_window()
+    return state
+
+def get_descendants_info(window):
+    result = []
+    for child in window.descendants():
+        info = child.get_properties()
+        descendants = get_descendants_info(child)  # Recursively get descendants of the current child
```
Comment (abrichr) [inline openadapt/window/_windows.py:55]: Please make sure we are not getting duplicates

File: openadapt/window/_windows.py (hunk 1 above)
Comment (jesicasusanto) [reply openadapt/window/_windows.py:34]: Done

File: openadapt/window/_windows.py (hunk 2 above)
Comment (jesicasusanto) [reply openadapt/window/_windows.py]: Done

**Comment by jesicasusanto** [comment]
Ready for review @abrichr

File: tests/openadapt/test_windows.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
+from openadapt.window._windows import get_active_window,get_descendants_info
+def test_check_duplicates():
+    active_window = get_active_window()
+    result = get_descendants_info(active_window)
+    unique_props = []
+    for item in result:
+        if isinstance(item, dict):
+            if item in unique_props:
+                import ipdb; ipdb.set_trace()
```
Comment (abrichr) [inline tests/openadapt/test_windows.py:9]: @jesicasusanto should we remove this?

Comment (jesicasusanto) [reply tests/openadapt/test_windows.py:9]: Yes I agree

File: openadapt/window/_windows.py [hunk 4]
```diff
@@ -1,30 +1,155 @@
...
+
+    Returns:
+        A nested dictionary containing the properties of each element and its children.
+        The dictionary includes a "children" key for each element, which holds the properties of its children.
+
+    Example:
+        element = Element()
+        properties = get_element_properties(element)
+        print(properties)
+        # Output: {'prop1': 'value1', 'prop2': 'value2', 'children': [{'prop1': 'child_value1', 'prop2': 'child_value2', 'children': []}]}
```
Comment (abrichr) [inline openadapt/window/_windows.py:116]: Can you please split this up (and other lines in this docstring) on multiple lines such that no line is greater than 80 characters?

File: openadapt/window/_windows.py [hunk 5]
```diff
@@ -1,30 +1,155 @@
...
+    if children:
+        properties["children"] = [get_element_properties(child) for child in children]
+
+    # Serialize the "rectangle" key
+    properties["rectangle"] = serialize_rect(properties["rectangle"])
+
+    return properties
+
+
+def serialize_rect(rect):
```
Comment (abrichr) [inline openadapt/window/_windows.py:131]: Is `serialize` the right word here? What about e.g. `dictify_rect`?

File: openadapt/window/_windows.py [hunk 6]
```diff
@@ -1,30 +1,155 @@
...
+    serialized = {
+        "left": rect.left,
+        "top": rect.top,
+        "right": rect.right,
+        "bottom": rect.bottom,
+    }
+    return serialized
+
+
+def main():
```
Comment (abrichr) [inline openadapt/window/_windows.py:141]: Maybe a docstring here would be useful, e.g. to indicate this is for testing?

**Review by abrichr — CHANGES_REQUESTED** [review]
Thank you @jesicasusanto ! Just a few small things. Can you also please show performance plots before and after these changes?

File: openadapt/window/_windows.py (hunk 4 above)
Comment (jesicasusanto) [reply openadapt/window/_windows.py]: Done @abrichr

File: openadapt/window/_windows.py (hunk 5 above)
Comment (jesicasusanto) [reply openadapt/window/_windows.py]: I think that sounds better and matches the purpose too. thank you! @abrichr

File: openadapt/window/_windows.py (hunk 6 above)
Comment (jesicasusanto) [reply openadapt/window/_windows.py:146]: @abrichr I agree, done!

**Comment by jesicasusanto** [comment]
Before :
image
After :
image

**Pull request merged**